cupy/logic/truth.py

```diff
@@ -57,3 +57,95 @@ def any(a, axis=None, out=None, keepdims=False):
 
     assert isinstance(a, cupy.ndarray)
     return a.any(axis=axis, out=out, keepdims=keepdims)
+
+
+def in1d(ar1, ar2, assume_unique=False, invert=False):
+    """
+    Tests whether each element of a 1-D array is also present in a second
+    array.
+
+    Returns a boolean array the same length as ``ar1`` that is ``True``
+    where an element of ``ar1`` is in ``ar2`` and ``False`` otherwise.
+
+    Args:
+        ar1 (cupy.ndarray): Input array.
+        ar2  (cupy.ndarray):
+            The values against which to test each value of ``ar1``.
+        assume_unique (bool, optional):
+            If ``True``, the input arrays are both assumed to be unique, which
+            can speed up the calculation. Default is ``False``.
+        invert(bool, optional):
+            If ``True``, the values in the returned array are inverted (that
+            is, ``False`` where an element of ``ar1`` is in ``ar2`` and
+            ``True`` otherwise). Default is ``False``.
+
+    Returns:
+        cupy.ndarray, bool: The values ``ar1[in1d]`` are in ``ar2``.
+
+    """
+    # TODO(UmashankarTriforce): Improve efficiency by avoiding loop over the
+    # first axis
+    # Ravel both arrays, behavior for the first array could be different
+    ar1 = ar1.ravel()
+    ar2 = ar2.ravel()
+
+    if invert:
+        mask = cupy.ones(len(ar1), dtype=bool)
+        for a in ar2:
+            mask &= (ar1 != a)
+    else:
+        mask = cupy.zeros(len(ar1), dtype=bool)
+        for a in ar2:
```

Comment:
After looking at the code carefully, I found the cost of this loop prohibitively large.
This loop incurs memory allocation of arrays as many as the number of elements in `ar2`.

I think the code can be simplified by using `broadcast_to` like below. How do you think?

```py
ar1 = ar1.ravel()
ar2 = ar2.ravel()
if ar1.size == 0 or ar2.size == 0:
    if invert:
        return cupy.ones(ar1.shape, dtype=cupy.bool_)
    else:
        return cupy.zeros(ar1.shape, dtype=cupy.bool_)

shape = (ar1.size, ar2.size)
ar1_broadcast = cupy.broadcast_to(ar1[..., cupy.newaxis], shape)
ar2_broadcast = cupy.broadcast_to(ar2, shape)
count = (ar1_broadcast == ar2_broadcast).sum(axis=1)
if invert:
    return count == 0
else:
    return count > 0
```

Reply:
I ran this code with tests. I think it looks good!